PMDMMOc/src/com/darkxell/client/graphics/floor/CameraVisibility.java

```diff
@@ -0,0 +1,59 @@
+package com.darkxell.client.graphics.floor;
+
+import java.util.HashSet;
+
+import com.darkxell.client.launchable.Persistence;
+import com.darkxell.client.state.dungeon.DungeonState;
+import com.darkxell.common.ai.visibility.Visibility;
+import com.darkxell.common.dungeon.data.FloorData;
+import com.darkxell.common.dungeon.floor.Floor;
+import com.darkxell.common.dungeon.floor.Tile;
+import com.darkxell.common.item.ItemEffects;
+import com.darkxell.common.item.ItemStack;
+import com.darkxell.common.pokemon.DungeonPokemon;
+
+public class CameraVisibility {
```

Comment:
Could it be possible to have a shared implementation with Visibility? Eg `#isItemVisible` is the exact same implementation atm.

Reply:
The CameraVisibility class can't extend Visibility because it isn't fixed to a single Pokemon. The camera Pokemon may change during a dungeon or may even be null. Also some rare cases have different behaviors, so we can't share the entire code. However there is currently no redundancy, I call methods from Visibility whenever it makes sense

Reply:
I meant as in visibility extends from class A, which CameraVisibility also extends from

Reply:
```java
/**
 * Wrapper class around the current camera's Visibility function. This is to easily handle null cases but may also have
 * different behaviors, because the camera may show things the Pokemon may not be able to see.
 */
public class CameraVisibility extends Visibility {
    public final DungeonState state;

    public CameraVisibility(DungeonState state) {
        this.state = state;
    }

    @Override
    public ai() {
        DungeonPokemon c = this.state.getCameraPokemon();
        return this.ai = c == null ? null : Persistence.floor.aiManager.getAI(c);
    }

    /** Updates AI to current camera AI, and returns if such an AI exists. */
    private boolean checkSetAI() {
        return this.ai() != null; // ai is a protected with getter, no longer public final
    }

    public boolean hasSeenTile(Tile tile) {
        return this.checkSetAI() && super().hasSeenTile(tile);
    }

    public boolean isItemTileVisible(Tile tile) {
        return this.checkSetAI() && super().isItemTileVisible(tile);
    }

    public boolean isMapVisible(DungeonPokemon pokemon) {
        return this.checkSetAI() && pokemon.player() == Persistence.player || super().isVisible(pokemon);
    }

    public boolean isVisible(DungeonPokemon pokemon) {
        if (Persistence.floor.data.shadows() == FloorData.NO_SHADOW)
            return true;
        return this.checkSetAI() && super().isVisible(pokemon);
    }
}
```

I meant something like this. I know it's a minor issue, but I feel that you should use inheritance because it shares so much of its implementation and its name. If most of the methods are just proxying to the current AI visibility, then it's probably a subclass.

Alternatively, you could do something like this:

```java
// Visibility.java, pseudocode
    public <T> static T checkCamera(DungeonState state, Visibility -> <T> callback) {
        DungeonPokemon c = this.state.getCameraPokemon();
        return c == null ? null : callback.accept(Persistence.floor.aiManager.getAI(c).visibility);
    }
// example usage, pseudocode
        Visibility.checkCamera(state, Visibility::isVisible);
```

CameraVisibility's verbosity and wrapper status just makes me not really like it, but ultimately it's a very minor issue.